pkg/virt-controller/network/ipamclaims.go

```diff
@@ -0,0 +1,205 @@
+	nonAbsentNetworks := filterNonAbsentNetworks(interfaces, networks)
+	networkToIPAMClaimParams, err := m.GetNetworkToIPAMClaimParams(namespace, vmiName, nonAbsentNetworks)
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	claims := composeIPAMClaims(namespace, ownerRef, networkToIPAMClaimParams)
+	if err := m.createIPAMClaims(namespace, claims); err != nil {
+		return fmt.Errorf("failed IPAMClaims creation for VMI %s: %w", vmiName, err)
+	}
+
+	return nil
+}
+
+func composeIPAMClaims(namespace string, ownerRef *v1.OwnerReference, networkToIPAMClaimParams map[string]IPAMClaimParams) []*ipamclaims.IPAMClaim {
+	claims := []*ipamclaims.IPAMClaim{}
+	for netName, ipamClaimParams := range networkToIPAMClaimParams {
+		claims = append(claims, composeIPAMClaim(
+			namespace,
+			*ownerRef,
+			ipamClaimParams,
+			namescheme.GenerateHashedInterfaceName(netName),
+		))
+	}
+
+	return claims
+}
+
+func (m *IPAMClaimsManager) createIPAMClaims(namespace string, claims []*ipamclaims.IPAMClaim) error {
+	for _, claim := range claims {
+		_, err := m.client.IPAMClaimsClient().K8sV1alpha1().IPAMClaims(namespace).Create(
+			context.Background(),
+			claim,
+			v1.CreateOptions{},
+		)
+
+		if err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create IPAMClaim: %w", err)
+			}
+
+			err = m.ensureValidIPAMClaimForVMI(namespace, claim.Name, claim.OwnerReferences[0].UID)
+			if err != nil {
+				return fmt.Errorf("failed validating IPAMClaim %s/%s: %w", namespace, claim.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *IPAMClaimsManager) ensureValidIPAMClaimForVMI(namespace string, claimName string, expectedOwnerUID types.UID) error {
+	currentClaim, err := m.client.IPAMClaimsClient().K8sV1alpha1().IPAMClaims(namespace).Get(
+		context.Background(),
+		claimName,
+		v1.GetOptions{},
+	)
+	if err != nil {
+		return fmt.Errorf("failed getting IPAMClaim: %w", err)
+	}
+
+	if len(currentClaim.OwnerReferences) != 1 || currentClaim.OwnerReferences[0].UID != expectedOwnerUID {
+		return fmt.Errorf("failed validating IPAMClaim, wrong IPAMClaim with the same name still exists")
+	}
+
+	return nil
+}
+
+func composeIPAMClaim(namespace string, ownerRef v1.OwnerReference, ipamClaimParams IPAMClaimParams, interfaceName string) *ipamclaims.IPAMClaim {
+	return &ipamclaims.IPAMClaim{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      ipamClaimParams.ClaimName,
+			Namespace: namespace,
+			OwnerReferences: []v1.OwnerReference{
+				ownerRef,
+			},
+		},
+		Spec: ipamclaims.IPAMClaimSpec{
+			Network:   ipamClaimParams.NetworkName,
+			Interface: interfaceName,
+		},
+	}
+}
+
+func (m *IPAMClaimsManager) GetNetworkToIPAMClaimParams(namespace string, vmiName string, networks []virtv1.Network) (map[string]IPAMClaimParams, error) {
+	nonMultusDefaultNetworks := vmispec.FilterMultusNonDefaultNetworks(networks)
```

Comment:
I think the filtering should be done prior calling to this method - same as for absent networks.

Reply:
but then every caller (there are 2 in production code, and 1 on unit tests) would need to call it
and it is less convenient / safe imho, unless you insist

Reply:
I don't understand what is the difference between the absent networks filtering to this one. Both should be done in the same place.

Reply:
Done